Phase 2

bjthorpe opened the pull request:
Cogs work thus far. Main features are fixes to #52 and #55 along with the creation of a new App called priority and an associated model. The priority calculations are invoked as: "python manage.py calculate-priority" at the moment it reads from the full output of  "sacct -P --start=2018-07-01 --allusers-o JobID,Account,Partition,CPUTimeRAW" which is stored as a local csv file under priority/fixtues/TEST_dump.csv which obviously will need changing to use SCP once we work out a route to/from sunbird.

Commit:
Removed a line that was for test output and fixed typo that totally brakes the logic

edbennett commented on cogs3/.template_env:
```diff
@@ -9,10 +9,10 @@ MAINTENANCE_MODE_IGNORE_ADMIN_SITE=False
+SECURE_CONTENT_TYPE_NOSNIFF=False
+SECURE_BROWSER_XSS_FILTER=False
+SESSION_COOKIE_SECURE=False
+CSRF_COOKIE_SECURE=False
```
This is useful for spinning up development instances, but it would be worth adding a comment in the README that the default implementation is insecure, and that these should be set to `True` for bringing up a production instance.

edbennett commented on funding/migrations/0001_initial.py:
```diff
@@ -6,6 +6,7 @@
 
 
 class Migration(migrations.Migration):
+    atomic = False
```
Do these migrations break if they are marked as atomic?

Commit:
removed erroneous migrate skeleton app

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        COGS_DB.to_sql('priority_slurm_priority',index=False,con=engine,if_exists='append')
+        engine.execute("SELECT * FROM priority_slurm_priority").fetchall()
+        
+def caculate_AP():
```
`caculate` should probably be `calculate`

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+    from project.models import Project, ProjectSystemAllocation
+    from funding.models import Attribution
+    from priority.models import slurm_priority
+    from datetime import date, timedelta
```
Most of these imports are present at the top of the file already—they can be removed here

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+    A = Project.objects.filter(code__in=on_sunbird)
+    priorities={}
+#query the Database for attributions associted to each project and calculate Ap based type of attribution
+    for B in A:
```
Please use more descriptive names than `A`, `B` and `C` here

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+    yesterday=date.today()-timedelta(1)
+    on_sunbird=[]
+    Sunbird=ProjectSystemAllocation.objects.exclude(system=1)
+    for I in Sunbird:
```
Please use a more descriptive name than `I` here

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+                    Ap=Ap+C.fundingsource.amount
+                else:
+                    Ap=Ap+10000
+                priorities.update({B.code:Ap})
```
The loop body here could be a method on the `Project` model. This would make it easier to access the current Ap level anywhere in the program, and would make the logic of this function easier to follow.

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+                if C.is_fundingsource:
+                    Ap=Ap+C.fundingsource.amount
+                else:
+                    Ap=Ap+10000
```
At some point we may want to add more types of attribution. Rather than considering the `else` case to be a publication, add an explicit check for a publication, and make the `else` case output a warning to a log file that an unknown attribution type was ignored in the priority calculation. (If we accidentally create objects of the wrong type, or add a type and forget to update this logic, this will then be clearer.)

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        slurm_priority.objects.create(Date=yesterday,Account='root',CPU_hours=0,GPU_hours=0,P_CPU=0,P_GPU=0,Qos=0)
+        Priority_DB = pd.DataFrame(list(slurm_priority.objects.all().values('Account','CPU_hours','GPU_hours','P_CPU','P_GPU','Qos')))
+        SCW=pd.merge(Priority_DB,Attpoints, on='Account', how='left').fillna(0) 
+    return SCW
```
Please use `autopep8` or similar on this file to keep the source as legible as possible.

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        #engine for reading/writing sql files
+        # sqlite://<nohostname>/<pathtofile>
+        engine = create_engine('sqlite:///db.sqlite3', echo=False)
+        df = pd.read_csv ('~/Documents/full_sacct_dump.csv',sep='|')
```
The location of the dump file should be supplied as an argument; this can be done by providing an `add_arguments(self, parser)` function, where the `parser` argument is an `argparse.ArgumentParser` object (if I'm reading the docs correctly)

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        SCW = pd.merge(SCW,gpu_sum, on='Account', how='outer').fillna(0)
+        #give all new entries an Ap of 50000
+        #SCW.update(default_Ap,overwrite=False,filter_func=new_data)
+        SCW.Ap.loc[SCW.Ap==0]=50000
```
This logic is already handled in `caculate_AP` - all projects start at 50k there, so no need to duplicate this logic here.

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        k=QOS_levels/np.log10(Max_Ap)
+        #check if previous run was prioritezed and if so add the time since last sactdump to total priortized time
+        P=SCW['Qos']>0
+        Prioritsed=SCW[P]
```
This should be `Prioritised` rather than `Prioritsed`

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        COGS_DB['Ap']=SCW['Ap']
+        COGS_DB['Date']=pd.date_range(start=Today, end=Today,periods=len(SCW.index))
+        #output Acoount and QOS to pipe seperated csv file 
+        Priority.to_csv(path_or_buf='~/Documents/Cogs3/cogs3/priority/fixtures/QOS_output.csv',sep='|',columns=['Account','QOS'],index=False)
```
Output location for QoS data to push to the cluster should be an argument

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        #seperate into gpu and not gpu
+        gpu=df[(df.Partition=="gpu")|(df.Partition=="xgpu")]
+        compute=df[(df.Partition!="gpu")|(df.Partition!="xgpu")]
+        compute.to_csv(path_or_buf='~/Documents/Cogs3/cogs3/compute.csv')
```
is this a test dump? If so, it should be optional, with the path given as an optional argument

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+#list all projects that are on sunbird
+    yesterday=date.today()-timedelta(1)
+    on_sunbird=[]
+    Sunbird=ProjectSystemAllocation.objects.exclude(system=1)
```
I'd like to avoid hardcoding "Sunbird" into the application where possible. Elsewhere we've done this by adding feature flags at an Institution level for features institutions want to opt into. In this case, we can apply the same AP calculation to everyone, since AP for non-Sunbird users will never be surfaced anywhere. (This also means that Cardiff and Bangor users temporarily running on Sunbird for whatever reason will be incorporated into the priority system, even if not being able to add attributions means that they will only get the initial 50k APs.)

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+    for B in A:
+        Ap=50000
+        if B.attributions==None:
+            priorities.update({B.code:Ap})
```
Isn't this the same as saying `priorities[B.code] = Ap`?

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+#query the Database for attributions associted to each project and calculate Ap based type of attribution
+    for B in A:
+        Ap=50000
+        if B.attributions==None:
```
This could be rephrased as:

```
if B.attributions:
    # Do the stuff in the lower block

priorities.update({B.code:Ap})
 ```

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        Today=date.today()
+        #engine for reading/writing sql files
+        # sqlite://<nohostname>/<pathtofile>
+        engine = create_engine('sqlite:///db.sqlite3', echo=False)
```
Does this still work if Django is configured to use a different database location, or different database manager?

edbennett commented on priority/management/commands/calculate-priority.py:
```diff
@@ -0,0 +1,121 @@
+        assert (not slurm_priority.objects.filter(Date__year=Today.year, Date__month=Today.month, Date__day=Today.day).exists()),'Data for today already exists in django database'
+        # Write data to django database
+        COGS_DB.to_sql('priority_slurm_priority',index=False,con=engine,if_exists='append')
+        engine.execute("SELECT * FROM priority_slurm_priority").fetchall()
```
This line seems to execute a query and then discard the results. Is this the intention?

edbennett reviewed (changes requested):
Thanks for your work on this, Ben, it's great progress. I've highlighted a couple of things that it'd be good to fix before we merge this and push it to Bangor. Let me know if you have any questions.

Commit:
changes to apication detail template to expose prioirty information to sunbird users

Commit:
I have updated the application detail template to expose priority infomation to the user using a new boolean variable needs_priority_workflow which I have added to the institution model. It defaults to false but is set to true for Swansea and Aber by the institutions.json file (this will need re-loading into the DB to set the changes). Also note that I have allowed external users to view the funding_workflow and priority_work flow information for testing as we don't have acess to shiboleth and therfore my admin user counts as external. This may need changing back if it is not the desired behaviour.

Commit:
Changes suggested by Ed for the pull request (Issue #58).
 Highlights include: Making Caculate_Ap a separate function in the
 project model so it can be used elsewhere, making the PD database
 update more robust, adding optional command-line arguments to
 specify input and output file locations and finally, making the
 code more readable.

bjthorpe reviewed (commented):
COMPANY_NAME='ACME Corporation'

SECRET_KEY=''

DEBUG=True

MAINTENANCE_MODE=False
MAINTENANCE_MODE_IGNORE_ADMIN_SITE=False

ALLOWED_HOSTS=["*"]

SECURE_CONTENT_TYPE_NOSNIFF=True
SECURE_BROWSER_XSS_FILTER=True
SESSION_COOKIE_SECURE=True
CSRF_COOKIE_SECURE=True
X_FRAME_OPTIONS='DENY'

SERVER_EMAIL=''
DEFAULT_SUPPORT_EMAIL=''
DEFAULT_FROM_EMAIL=''
DEFAULT_BCC_EMAIL=''
DEFAULT_CONTACT_EMAIL=''

EMAIL_BACKEND= ''
EMAIL_FILE_PATH=''
EMAIL_HOST=''
EMAIL_HOST_USER=''
EMAIL_HOST_PASSWORD=''
EMAIL_PORT=''
EMAIL_USE_TLS=True

RQ_HOST=''
RQ_PORT=''
RQ_DB=''
RQ_PASSWORD=''
RQ_DEFAULT_TIMEOUT=360

OPENLDAP_HOST=''
OPENLDAP_JWT_KEY=''
OPENLDAP_JWT_ISSUER=''
OPENLDAP_JWT_AUDIENCE=''
OPENLDAP_JWT_ALGORITHM=''

SHIBBOLETH_IDENTITY_PROVIDER_LOGIN=''
SHIBBOLETH_IDENTITY_PROVIDER_LOGOUT=''
